15_tictactoe/go/io/io.go

```diff
@@ -2,40 +2,35 @@ package io
 
 import (
 	"fmt"
+	"slices"
 	"strconv"
 )
 
-/*
-Input validation:
-Create loop
-Break loop if input is an integer, else keep looping
-Break loop if input is on a free cell, else keep looping
-*/
-
-func Read(prompt string) int {
+func Read(prompt string, freeCellSlice []int) int {
 	var input string
 	for {
 		fmt.Printf("%s > ", prompt)
 		fmt.Scan(&input)
 
-		if valid(input) {
+		if valid(input, freeCellSlice) {
 			validInput, _ := strconv.Atoi(input)
 			return validInput
 		}
 	}
 }
 
-func valid(input string) bool {
+func valid(input string, freeCellSlice []int) bool {
 	intInput, err := strconv.Atoi(input)
+
 	if err != nil {
 		fmt.Print("Invalid input! Only single digit numbers allowed!\n")
 		return false
 	}
-	if intInput >= 0 && intInput <= 8 {
+
+	if slices.Contains(freeCellSlice, intInput) {
 		return true
 	}
 
-	fmt.Println("Print invalid input", intInput)
+	fmt.Printf("'%v' is already taken! Try another choice\n", intInput)
```

Comment:
```suggestion
	fmt.Printf("'%d' is already taken! Try another choice\n", intInput)
```

`%v` works in most cases, but `%d` is better suited for integer/numbers: https://en.wikipedia.org/wiki/Printf#Type_field

Reply:
Done. I guess go as precise as possible then.